lib/srv/sess.go

```diff
@@ -1494,6 +1558,103 @@ func (s *session) checkIfFileTransferApproved(req *fileTransferRequest) (bool, e
 	return isApproved, nil
 }
 
+// newFileTransferRequest takes FileTransferParams and creates a new fileTransferRequest struct
+func (s *session) newFileTransferRequest(params *rsession.FileTransferParams) *fileTransferRequest {
+	return &fileTransferRequest{
+		id:        uuid.New().String(),
+		requester: params.Requester,
+		location:  params.Location,
+		filename:  params.Filename,
+		size:      params.Size,
+		direction: params.Direction,
+		approvers: make(map[string]*party),
+	}
+}
+
+// addFileTransferRequest will create a new file transfer request and add it to the current session's fileTransferRequests map
+// and broadcast the appropriate string to the session.
+func (s *session) addFileTransferRequest(params *rsession.FileTransferParams, scx *ServerContext) *fileTransferRequest {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	req := s.newFileTransferRequest(params)
+	s.fileTransferRequests[req.id] = req
+	if params.Direction == "download" {
+		s.BroadcastMessage("User %s would like to download: %s", params.Requester, params.Location)
+	} else {
+		s.BroadcastMessage("User %s would like to upload %s to: %s", params.Requester, params.Filename, params.Location)
+	}
+
+	return req
+}
+
+// approveFileTransferRequest will add the approver to the approvers map of a file transfer request and notify the members
+// of the session if the updated approvers map would fulfill the moderated policy.
+func (s *session) approveFileTransferRequest(params *rsession.FileTransferParams, scx *ServerContext) (*fileTransferRequest, error) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	fileTransferReq := s.fileTransferRequests[params.RequestID]
+	if fileTransferReq == nil {
+		return nil, trace.NotFound("File Transfer Request %s not found", params.RequestID)
+	}
+
+	var approver *party
+	for _, p := range s.parties {
+		if p.ctx.ID() == scx.ID() {
+			approver = p
+		}
+	}
+	if approver == nil {
+		return nil, trace.AccessDenied("Cannot approve file transfer requests if not in the current moderated session")
+	}
+
+	fileTransferReq.approvers[approver.user] = approver
+	s.BroadcastMessage("%s approved file transfer request %s", scx.Identity.TeleportUser, fileTransferReq.id)
+
+	// check if policy is fulfilled
+	approved, err := s.checkIfFileTransferApproved(fileTransferReq)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	var eventType FileTransferRequestEvent
+	if approved {
+		eventType = FileTransferApproved
+	} else {
+		eventType = FileTransferUpdate
+	}
+
+	s.registry.NotifyFileTransferRequest(fileTransferReq, eventType, scx)
+
+	return fileTransferReq, nil
+}
+
+// denyFileTransferRequest will deny a file transfer request and remove it from the current session's file transfer requests map.
+// A file transfer request does not persist after deny, so there is no "denied" state. Deny in this case is synonymous with delete
+// with the addition of checking for a valid denier.
```

Comment:
So what happens in this scenario?

- 2 moderator approvals are required for transfer
- 3 moderators are watching the session
- moderator A approves the transfer
- moderator B denies the transfer

Does the request immediately get dropped, or does moderator C still have a chance to approve?

Reply:
Request will immediately be dropped. Similar to access requests, if moderated A approved and B denied, C won't be able to answer as the result is already in a `denied` state.